### ISSUE TITLE
Hiding-showing proofs in coqide

### ISSUE THREAD

**Issue by MdeLv** [coq/coq#12088]
The documentation talks about a Navigation / Hide proof menu item within coqide:
<https://coq.inria.fr/refman/practical-tools/coqide.html#proof-folding>
<https://github.com/coq/coq/blob/master/doc/sphinx/practical-tools/coqide.rst#proof-folding>

I can't find this must-have feature (in the files in ~/.config/coq/ or in coqide preferences)
How can we enable it?  
Thanks

coq.8.11.1

**Comment by Zimmi48** [comment]
This feature appears to have been removed since 8.5. It's not likely that it will be reintroduced although @coq/coqide-maintainers might say. In any case, that's a documentation bug to fix.

**Comment by Zimmi48** [comment]
Leaving this open as enhancement request.

**Comment by MdeLv** [comment]
Is this hide/show proof feature available within Emacs/Proof General?

BTW, where can one find some reliable Coq/Emacs installation instructions? (Proof General manual seems not enough; at least for me).

Thanks.

**Comment by Zimmi48** [comment]
Yes, it is available in Proof General.

What do you mean by not enough?

cc @cpitclaudel @erikmd @Matafou

**Comment by erikmd** [comment]
> where can one find some reliable Coq/Emacs installation instructions? (Proof General manual seems not enough; at least for me)

which OS do you have?

indeed, the PG manual itself is a bit terse in this regard for the moment; but you may want to have a look at these two comprehensive installation tutorials:

* an Emacs+OCaml+Coq installation tutorial (in French) for [GNU/Linux](https://github.com/erikmd/tapfa-init.el), [Windows 10](https://github.com/erikmd/tapfa-init-win64.el), or [macOS](https://github.com/erikmd/tapfa-init-macos.el) − but feel free to skip the unneeded step `opam install -y merlin utop learn-ocaml-client`
* the [Coq wiki itself](https://github.com/coq/coq/wiki#coq-installation) which notably suggests other possible setups if you have a [Windows environment](https://github.com/coq/coq/wiki/Installation%20of%20Coq%20on%20Windows)

**Comment by MdeLv** [comment]
@Zimmi48: I mean "PG manual seems not enough to get up and running with Emacs/Proof General/Coq (without looking under the hood, which requires knowledge about Emacs)."

@erikmd: I use Debian. Your .emacs file is great for Coq and even for OCaml. Thanks.

I just had a message from Emacs (26.1) while contacting packages repository at port 443. I may have answered incorrectly (no | yes ?) to a simple but ambiguous security message,
So I did: M-x load-file ~/.emacs and the installation started and succeeded (because I know load-file. As a total Emacs beginner I could get stuck, hence my feedback). 

Proof General shortcuts are available.
However, I can't hide/show blocks as expected (menu items are displayed). 
I try to place the cursor somewhere between Proof and Qed, or to select a region including "Proof." and "Qed.", then select 'Hide block' menu item, but nothing happens.
How do you use these hide/show (proof or comments) features? (what is a block and do you select it?) 

BTW, selected text is not highlighted with this .emacs file. This made the hide/show features more problematic, so I did: M-x customize face region and selected, say, gold1 as Background. Fixed.
Here is the generated code that you may include in your .emacs file if you like:
```
(custom-set-faces
 '(proof-locked-face ((t (:background "#add8e6"))))
 '(region ((t (:background "gold1" :distant-foreground "gtk_selection_fg_color")))))
```

**Comment by cpitclaudel** [comment]
Btw, you might be interested in company-coq.  It has proof folding on `C-c C-/`.

![](https://raw.githubusercontent.com/cpitclaudel/company-coq/master/img/folding.gif)

**Comment by erikmd** [comment]
Hi @MdeLv

> I just had a message from Emacs (26.1) while contacting packages repository at port 443. I may have answered incorrectly (no | yes ?) to a simple but ambiguous security message,

very probably, you've hit the following Emacs/Debian bug :

https://debbugs.gnu.org/cgi/bugreport.cgi?bug=34341

which is solved in Emacs 26.3 and 27.1, but according to that bugreport a working workaround for Emacs<26.3 consists in adding at the beginning of your .emacs:

```elisp
(setq gnutls-algorithm-priority "NORMAL:-VERS-TLS1.3")
```

**Comment by erikmd** [comment]
> I try to place the cursor somewhere between Proof and Qed, or to select a region including "Proof." and "Qed.", then select 'Hide block' menu item, but nothing happens.

I did not test this feature right away, but normally this (plain PG's) Proof hiding feature should work only if you already processed the proof at stake (e.g. with `C-c C-RET` up to Qed).

Anyway as suggested by @cpitclaudel, when using Coq-ProofGeneral, company-coq is a must!

**Comment by MdeLv** [comment]
@cpitclaudel: the .emacs from @erikmd includes company-mode.
It just need to be permanently enabled, as suggested after first C-c C-/ .

Your [Quick Start Guide](https://github.com/cpitclaudel/company-coq#quick-start-guide) is nicely illustrated and helpful. 
I found there how to unfold proofs with C-c C-\ (before that, I was stuck with hidden proofs, as I didn't see immediately the [...] that denotes a folded proof).

Can you show me where to look to customize this fold/unfold feature as follows?
```
Lemma Transitive_eq : forall A (x y z: A), x = y -> y = z -> x = z.
Proof [...]
```
It's the same except that 'Proof.' is left visible and '[...]' placed just after it. It's a chance to explore Emacs a little bit more.

BTW, the company/'Help for minor mode' on bottom status line opens a small documentation.
What do you think about adding the 'M-x company-coq-tutorial RET' invitation and a link to https://github.com/cpitclaudel/company-coq at the first lines of this immediate help page?

**Comment by MdeLv** [comment]
@erikmd:
As of today, Debian Buster has still [Emacs 26.1](https://packages.debian.org/buster/emacs).
I deleted the installed packages and tried again with your .emacs file and this TLS patch.
Test: OK (no more TLS error; just a small error while verifying signature archive-contents.sig).

The Hide/Show menu items (from PG) seem to have no effect even if a proof is processed.
Anyway, company-coq does the job.
Thanks.

**Comment by cpitclaudel** [comment]
> Can you show me where to look to customize this fold/unfold feature as follows?

Yup, you'd want to change company-coq-outline-regexp.  Teh reason I didn't do it by default is that Proof is an optional keyword.

**Comment by Matafou** [comment]
The feature in PG was bugged indeed. It is hopefully fixed now on pg master. But it is less precise than in company-coq: only the whole proof can be hidden, no bullet-grained hiding.

**Comment by erikmd** [comment]
Thanks @Matafou and @MdeLv !
